biochatter/api_agent.py

```diff
@@ -0,0 +1,441 @@
+
+        Args:
+            file_path (str): The path to the file.
+            n (int): The number of lines to read.
+
+        Returns:
+            str: The first n lines from the file as a string.
+
+        Raises:
+            FileNotFoundError: If the file is not found.
+            Exception: If any other error occurs during file reading.
+
+        """
+        try:
+            with open(file_path, "r") as file:
+                lines = []
+                for i in range(n):
+                    line = file.readline()
+                    if not line:
+                        break
+                    lines.append(line.strip())
+                # to test:
+                # more efficient with \n or without?
+                return "\n".join(lines)
+        except FileNotFoundError:
+            return "The file was not found."
+        except Exception as e:
+            return f"An error occurred: {e}"
+
+    def answer_extraction(self, question: str, file_path: str, n: int) -> str:
+        """
+        Function to extract the answer from the BLAST results.
+
+        Args:
+            question (str): The question to be answered.
+            file_path (str): The path to the BLAST results file.
+            n (int): The number of lines to read from the file.
+
+        Returns:
+            str: The extracted answer from the BLAST results.
+
+        """
+        prompt = ChatPromptTemplate.from_messages(
+            [
+                (
+                    "system",
+                    "You are a world class molecular biologist who knows everything about NCBI and BLAST results.",
+                ),
+                ("user", "{input}"),
+            ]
+        )
+
+        context = self.read_first_n_lines(file_path, n)
+        BLAST_file_answer_extractor_prompt = f"""
+                You have to answer the question: {question} in a clear and concise manner. Be factual!\n\
+                If you are asked what organism a specific sequence belongs to, check the 'Hit_def' fields. If you find a synthetic construct or predicted entry, move to the next one and look for an organism name.\n\
+                Try to use the hits with the best identity score to answer the question. If it is not possible, move to the next one.\n\
+                Be clear, and if organism names are present in ANY of the results, please include them in the answer. Do not make up information and mention how relevant the found information is based on the identity scores.\n\
+                Use the same reasoning for any potential BLAST results. If you find information that is manually curated, please use it and state it. You may also state other results, but always include the context.\n\
+                Based on the information given here:\n\
+                {context}
+                """
+        output_parser = StrOutputParser()
+        chain = prompt | llm | output_parser
+        BLAST_answer = chain.invoke(
```

Comment:
using the langchain idiomatic style is constraining, in this case probably unnecessarily. would refactor to be a simple query built from the parts